[WIP] TAAR Ensemble

florian opened the pull request:
**WIP: This is not a final version of the notebook**. It's mostly to give a preview of what I'm doing because it's probably a bad idea to push a huge notebook that nobody ever saw before on my last day.

This notebook contains an implementation of the linear stacking approach that's [described on Bugzilla](https://bugzilla.mozilla.org/show_bug.cgi?id=1400186). The following things are part of this notebook:

- Each recommender now contains a function `get_weighted_recommendations` for retrieving confidence scores
- Spark code for finding a subset of users to train on, for masking addons and for computing all features
- An optimization process for finding good weights
- A comparison of the previous recommendation manager

The optimization results are not very stable right now, but the ensemble generally performs much better than the previous recommendation manager on MAP. A better baseline for comparison is just the locale recommender. Here the ensemble has a score that's about 2% better.

florian commented:
The following things are still missing as of now:
- Comparing results on a validation/test set. Of course this means that there's a chance that the current ensemble model is overfitting and that the comparison is not fair
- A more stable optimization process (e.g. with grid search and by using some intuition on how the error surface looks like)
- Better feature normalization and preprocessing

mlopatka commented on Ensemble-Model.md:
```diff
@@ -0,0 +1,1066 @@
+
+### Collaborative Recommender
+
+Use the scores computed internally
```
It would be useful to add a comment explaining what the confidence number actually represents in terms of the internal model used by this recommender.

mlopatka commented on Ensemble-Model.md:
```diff
@@ -0,0 +1,1066 @@
+        indices = (-lrs_from_scores).argsort()
+        return lrs_from_scores[indices], indices
+
+    def get_weighted_recommendations(self, client_data):
```
Likewise here, what is the _meaning_ of the confidence value for this module.

florian replied:
I added some more explanations.

mlopatka commented on Ensemble-Model.md:
```diff
@@ -0,0 +1,1066 @@
+            
+        self.top_addons_by_locale = defaultdict(lambda: defaultdict(int), top_addons_by_locale)
+        
+    def get_weighted_recommendations(self, client_data):
```
is this normalized per locale or per total number of addons in all locales?
Add an explanation as to what the confidence means for this module.

florian replied:
It's normalized per locale. I just added more explanations to the notebook.

mlopatka commented on Ensemble-Model.md:
````diff
@@ -0,0 +1,1066 @@
+
+
+```python
+useful_clients = completed_rdd.filter(lambda client: len(client['installed_addons']) >= 1).cache()
````
How many `useful_clients` do we have left if this condition is changed to:
`len(client['installed_addons']) >= 2)`
`len(client['installed_addons']) >= 3)`
`len(client['installed_addons']) >= 4)`

florian replied:
Very very few clients. There's a table with exact numbers at the bottom of the [evaluation notebook](https://github.com/florian/taar-prototyping/blob/master/evaluation.ipynb). I'll also link this notebook in the ensemble notebook.

- `len(client['installed_addons']) >= 2)` => ~0.01704%
- `len(client['installed_addons']) >= 3)` => ~0.00264%
- `len(client['installed_addons']) >= 4)` => ~0.00122%

mlopatka commented on Ensemble-Model.md:
````diff
@@ -0,0 +1,1066 @@
+
+```python
+def preprocess_locale_scores(scores):
+    return np.sqrt(np.sqrt(scores))
````
Why is this necessary if the others are normalized using min_max scaling? 
If the objective is to have a value between 0.0 and 1.0 then would a sum normalization (per locale) not be better?

florian replied:
>  If the objective is to have a value between 0.0 and 1.0 then would a sum normalization (per locale) not be better?

I'm already applying such a normalization before, in the locale_job notebook.

The intuition behind another transformation (something like sqrt or log) is that some addons otherwise have extremely high confidence scores, which would lead to these addons (like µBlock) always being recommended. With a sqrt/log transformation, these addons still have an advantage but it's not as extreme anymore.

I'll run some more tests tomorrow to see if this is actually useful or if it doesn't make much of a difference.

mlopatka commented on Ensemble-Model.md:
````diff
@@ -0,0 +1,1066 @@
+
+
+```python
+max_feature_values = X_unnormalized.map(lambda (_, features): np.max(features, axis=0)).reduce(np.maximum)
````
Are the different confidence values all linearly scalable> I.e. does a min_max transformation for the collaborative recommender mean the same thing as a min_max transformation for the similarity recommender?

Are the relationships between addon recommendation confidence comparable after the transformation?

florian replied:
> I.e. does a min_max transformation for the collaborative recommender mean the same thing as a min_max transformation for the similarity recommender?

Nope, unfortunately not. My plan was to find some transformations so that this is nearly the case, but I didn't get to that.

mlopatka commented on Ensemble-Model.md:
````diff
@@ -0,0 +1,1066 @@
+
+```python
+def cost(weights):
+    return 1 - X.map(lambda (client_data, features): get_weighted_recommendations(client_data, features, weights))\
````
nit: break this line up further for readability please.

mlopatka commented on Ensemble-Model.md:
````diff
@@ -0,0 +1,1066 @@
+
+
+```python
+training, validation, test = useful_clients.randomSplit([0.8, 0.1, 0.1])
````
perhaps try something closer to [0.6, 0.2, 0.2] for a larger number of points evaluated against the cost function.

florian replied:
I'm not using the validation set anyways right now, so I changed it to `[0.8, 0, 0.2]` for now. 

Originally, I was planing to use a validation set for things like early stopping. I didn't really have enough time for that though and it doesn't seem like overfitting by training for too long is something that happens here anyways.

mlopatka commented on Ensemble-Model.md:
```diff
@@ -0,0 +1,1066 @@
+num_features = len(recommenders)
+x0 = get_initial_guess(num_features)
+print "Initial guess:", x0
+minimize(verbose_cost, x0, method="COBYLA", bounds=[(0, None)] * num_features, tol=1e-10)
```
Please explain what the COBYLA method is and why you selected this strategy

mlopatka commented on Ensemble-Model.md:
```diff
@@ -0,0 +1,1066 @@
+
+
+
+However, this comparison is a little bit unfair. The locale recommender is generally extremely useful and can be used as a better baseline. With this ordering (where nearly only the locale recommender is queried), we get a much more comparable result. The results are now in the same ballpark and the ensemble is better by around 2%.
```
This is also a bit circular logic as the locale recommender uses segregates counts of addons installed per locale predict the addons to recommend per locale :)

The same data source is used to generate the counts that locale_recommender uses as is used here as an evaluation set. Therefore the generalizability only holds so long as locale-specific trends are static. Also to be noted is that the locale recommender is biased towards the most popular addons by virtue of its construction so the performance over `useful_clients` with a larger number of installed addons. This may be especially relevant when new users arrive with popular addons already installed but expect interesting recommendations.

Dexterp37 commented on Ensemble-Model.md:
```diff
@@ -0,0 +1,1066 @@
+    
+    addons = client['installed_addons'] or []
+    client['installed_addons'] = get_whitelisted_addons(addons)
+    client['disabled_addons_ids'] = addons
```
Why are you setting `disabled_addons_ids` to the `installed_addons`? I know that we don't have the disabled add-on info in the Longitudinal dataset. However, only putting legacy addons would be more appropriate here. Could you expand your reasoning here?

florian replied:
I'm just using that dictionary key because it is the one that the [`LegacyRecommender`](https://github.com/mozilla/taar/blob/master/taar/recommenders/legacy_recommender.py#L44) expects. The LegacyRecommender only takes into account addons in its replacement list anyways, so adding another filter for legacy addons on my part wouldn't make a difference.

Dexterp37 commented on Ensemble-Model.md:
````diff
@@ -0,0 +1,1066 @@
+
+
+```python
+def get_initial_guess(n):
````
Please leave one strategy and disable/comment the other. When running this notebook, the random initialization will always prevail otherwise. Let's make the choice explicit.

Dexterp37 commented on Ensemble-Model.md:
```diff
@@ -0,0 +1,1066 @@
+
+    
+
+    KeyboardInterruptTraceback (most recent call last)
```
Looks like this method was interrupted? Why is this output here? Is this optimization being used?

florian replied:
I just stopped it manually (interrupt kernel / ctrl+c) because I saw that the optimization didn't improve much anymore and I wanted to execute other things. :)

I fully ran it now, so that the notebooks looks a bit cleaner.

florian commented:
I pushed some updates addressing the comments. I also added a new section about some more manual optimization.

The changes to the existing TAAR recommenders that are used in this notebook can be found in the [ensemble_changes branch of the florian/taar repository](https://github.com/florian/taar/tree/ensemble_changes).
